File: p2p_extension/signaling.js
let socket;
let myId;

function connectSignaling(id) {
  myId = id;
  socket = new WebSocket("ws://localhost:3000");

  socket.onopen = () => {
    socket.send(JSON.stringify({ join: true, userId: myId }));
  };

  socket.onmessage = e => {
    const data = JSON.parse(e.data);
    handleSignal(data);
  };
}

function sendSignal(to, type, payload) {
  socket.send(JSON.stringify({
    from: myId,
    to,
    type,
    payload
  }));
}

File: poc_app_extension/lib/signaling.dart
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';

class Signaling {
  final String url;
  late WebSocketChannel _channel;
  late String myId;

  Function(Map<String, dynamic>)? onMessage;

  Signaling(this.url);

  void connect(String userId) {
    myId = userId;
    _channel = WebSocketChannel.connect(Uri.parse(url));

    _channel.sink.add(jsonEncode({
      "join": true,
      "userId": myId,
    }));

    _channel.stream.listen((message) {
      onMessage?.call(jsonDecode(message));
    });
  }

  void send({
    required String to,
    required String type,
    required dynamic payload,
  }) {
    _channel.sink.add(jsonEncode({
      "from": myId,
      "to": to,
      "type": type,
      "payload": payload,
    }));
  }

  void dispose() {
    _channel.sink.close();
  }
}
